### DISCUSSION TITLE
[mini.pick] Augment `Pick buffers` with hotkeys

### DISCUSSION THREAD

**Discussion by abeldekat** [nvim-mini/mini.nvim#1096] [Show and tell]
https://github.com/user-attachments/assets/14f9596a-c18d-4970-8770-252a87085839

My workflow regarding buffer navigation is centered around [grapple.nvim](https://github.com/cbochs/grapple.nvim). I do use `Pick buffers` often. Usually, its `query` contains more than one character, followed by `enter`. It would be nice to only type one single "hotkey", in `Jump2d` fashion.

Trying my hand on this requirement has been on my `todo-list` for some time. Recently, plugin [snipe.nvim](https://github.com/leath-dub/snipe.nvim) was announced on reddit, reminding me on that "todo".

I tried to incorporate `Jump2d` into `Pick` in various ways. No cigar unfortunately. The main `Pick` window disappears before `Jump2d` has a chance to kick-in.

In my opinion, there exists a sweet spot between the amount of opened buffers and the benefit accomplished by using a "hotkey". When the amount of buffers is considerable, the best approach is to narrow down the list using default `Pick buffers`. Plugin `snipe.nvim` uses multiple pages in that situation, a less optimal approach.

I took the code from [Pick buffers](https://github.com/echasnovski/mini.nvim/blob/10426fc5bd38397c9a74d80cd4a6c55243fc78de/lua/mini/pick.lua#L1427) and implemented [Pick labeled_buffers](https://github.com/abeldekat/nvim_pde/blob/5ac3e75aeb7331bb8674b9d58d5845c0876ac86f/lua/ak/config/editor/mini_pick.lua#L63)

Algorithm:

If the amount of labels can be used to add "hotkeys" to **all** buffer items, the labels are highlighted. Otherwise, the labels are dimmed, but they are still available for faster navigation. `Pick` operates as usual in that case.

The straightforward matching algorithm provided by `Pick` simplified the implementation.
Just add a unique letter before the text representing the buffer, and done.
For now, I centered the `pick` window, to more closely resemble the centered `grapple` window. Also, as the text representing the buffer can be lengthy, I _surrounded_ the text with the labels.

The code:

```lua
Pick.registry.labeled_buffers = function(_, _) -- local_opts, opts
  local letters = "abcdefghijklmnopqrstuvwxyz" -- more items than letters: no hotkey
  local buffer_items = (function() -- see Pick.builtin.buffers
    local add_label = function(text, idx)
      local letter = string.sub(letters, idx, idx)
      local label = letter == "" and "[ ]" or ("[" .. letter .. "]")
      return string.format("%s %s %s", label, text, label) -- the first label does the trick!
    end
    local buffers_output = vim.api.nvim_exec2("buffers", { output = true }).output
    local buffer_items, cur_buf_id, include_current = {}, vim.api.nvim_get_current_buf(), false
    for i, l in ipairs(vim.split(buffers_output, "\n")) do
      local buf_str, name = l:match("^%s*%d+"), l:match('"(.*)"')
      local buf_id = tonumber(buf_str)
      local item = { text = add_label(name, i), bufnr = buf_id } -- additionally add label
      if buf_id ~= cur_buf_id or include_current then table.insert(buffer_items, item) end
    end
    return buffer_items
  end)()
  local window = true
      and { -- centered window for consistency with grapple
        config = function() -- copied from :h MiniPick
          local height = math.floor(0.618 * vim.o.lines)
          local width = math.floor(0.618 * vim.o.columns)
          return {
            anchor = "NW",
            height = height,
            width = width,
            row = math.floor(0.5 * (vim.o.lines - height)),
            col = math.floor(0.5 * (vim.o.columns - width)),
          }
        end,
      }
    or nil
  local show_icons = true
  local icon_length = show_icons and 5 or 0
  local use_hotkey = #buffer_items <= string.len(letters)
  local hl = use_hotkey and "MiniPickMatchRanges" or "Comment"

  local show = function(buf_id, items, query)
    local find_label = function(item, find_pos)
      local start, _ = string.find(item.text, "[", find_pos, true)
      start = start and (start - 1) + icon_length or start
      return start
    end
    MiniPick.default_show(buf_id, items, query, { show_icons = show_icons })

    if use_hotkey and query and #query == 1 then
      local submit = vim.api.nvim_replace_termcodes("<cr>", true, false, true)
      vim.api.nvim_feedkeys(submit, "n", false) -- hotkey
    end
    for i = 1, #items do
      local start = find_label(items[i], 1) -- first label occurrence
      if start then vim.api.nvim_buf_add_highlight(buf_id, 0, hl, i - 1, start, start + 3) end
      start = find_label(items[i], 2) -- second label occurrence
      if start then vim.api.nvim_buf_add_highlight(buf_id, 0, hl, i - 1, start, -1) end
    end
  end
  return MiniPick.start({ window = window, source = { name = "Labeled_buffers", items = buffer_items, show = show } })
end
```

**Comment by echasnovski** [comment]
That's a nice idea!

To be perfectly honest, I'd say that code look a bit too complex for what it eventually achieves. I especially wouldn't recommend the `x = (function() ... end)()` approach as it is not as straightforward to read as without it (or with `do ... end` if absolutely necessary).

My mind immediately went in the direction of making a more general `make_labeled_picker()`. Which takes a picker source and modifies it to be "labeled":
- Wraps each item (directly or callable's output) in a table with a dedicated `label`, `item`, and `text` fields.
- Overrides `match` to perform a single (or several until necessary) key match.
- Possibly overrides `show` to have labels a special highlight (not sure if necessary).

**Reply by abeldekat** [reply]
Thanks for the feedback! I am working on the changes.

**Reply by abeldekat** [reply]
See [this](https://github.com/echasnovski/mini.nvim/discussions/1109) discussion

**Comment by leath-dub** [comment]
Good work on this, I do think a hybrid would be interesting to investigate although i personally never really exceed 20 buffers

**Reply by abeldekat** [reply]
Thanks!
